File: go/abi.go
package omne

import (
	"encoding/binary"
	"encoding/hex"
	"fmt"
	"math/big"
)

// ArgType discriminants matching the Rust ABI codec.
type ArgType byte

const (
	ArgTypeU32     ArgType = 0x01
	ArgTypeU64     ArgType = 0x02
	ArgTypeI32     ArgType = 0x03
	ArgTypeI64     ArgType = 0x04
	ArgTypeString  ArgType = 0x05
	ArgTypeBytes   ArgType = 0x06
	ArgTypeBool    ArgType = 0x07
	ArgTypeAddress ArgType = 0x08
	ArgTypeU128    ArgType = 0x09
)

// AbiArgument is a single typed argument in the Omne ABI encoding.
type AbiArgument struct {
	Type ArgType
	Data []byte
}

// Arg* builders produce typed ABI arguments. Integers are BIG-endian (the
// codec), distinct from the little-endian transaction-hash preimage.

func ArgU32(v uint32) AbiArgument {
	b := make([]byte, 4)
	binary.BigEndian.PutUint32(b, v)
	return AbiArgument{ArgTypeU32, b}
}

func ArgU64(v uint64) AbiArgument {
	b := make([]byte, 8)
	binary.BigEndian.PutUint64(b, v)
	return AbiArgument{ArgTypeU64, b}
}

func ArgI32(v int32) AbiArgument {
	b := make([]byte, 4)
	binary.BigEndian.PutUint32(b, uint32(v))
	return AbiArgument{ArgTypeI32, b}
}

func ArgI64(v int64) AbiArgument {
	b := make([]byte, 8)
	binary.BigEndian.PutUint64(b, uint64(v))
	return AbiArgument{ArgTypeI64, b}
}

// ArgU128 encodes a non-negative big.Int as a 16-byte big-endian value.
func ArgU128(v *big.Int) AbiArgument {
	b := make([]byte, 16)
	v.FillBytes(b) // big-endian, left-padded; panics if v < 0 or > 16 bytes
	return AbiArgument{ArgTypeU128, b}
}

// ArgU128FromUint is a convenience for u128 values that fit in a uint64.
func ArgU128FromUint(v uint64) AbiArgument {
	return ArgU128(new(big.Int).SetUint64(v))
}

func ArgString(s string) AbiArgument { return AbiArgument{ArgTypeString, []byte(s)} }

func ArgBytes(b []byte) AbiArgument { return AbiArgument{ArgTypeBytes, b} }

func ArgBool(v bool) AbiArgument {
	if v {
		return AbiArgument{ArgTypeBool, []byte{0x01}}
	}
	return AbiArgument{ArgTypeBool, []byte{0x00}}
}

// ArgAddress encodes an om1z address as a 32-byte Address argument.
func ArgAddress(omneAddress string) (AbiArgument, error) {
	payload, err := ParseAddress(omneAddress)
	if err != nil {
		return AbiArgument{}, err
	}
	if len(payload) != 32 {
		return AbiArgument{}, fmt.Errorf("address must be 32 bytes, got %d", len(payload))
	}
	return AbiArgument{ArgTypeAddress, payload}, nil
}

// EncodeContractCall encodes a method call into the Omne ABI wire format
// (bare hex): MAGIC "OMNE" | VERSION 0x01 | method_len u16 BE | method |
// arg_count u16 BE | [type u8 | data_len u32 BE | data]*.
func EncodeContractCall(method string, args []AbiArgument) (string, error) {
	mb := []byte(method)
	if len(mb) > 256 {
		return "", fmt.Errorf("method name exceeds 256 byte limit")
	}
	if len(args) > 64 {
		return "", fmt.Errorf("argument count %d exceeds maximum of 64", len(args))
	}
	buf := []byte{0x4f, 0x4d, 0x4e, 0x45, 0x01} // "OMNE" + version
	var u16 [2]byte
	binary.BigEndian.PutUint16(u16[:], uint16(len(mb)))
	buf = append(buf, u16[:]...)
	buf = append(buf, mb...)
	binary.BigEndian.PutUint16(u16[:], uint16(len(args)))
	buf = append(buf, u16[:]...)
	for _, a := range args {
		buf = append(buf, byte(a.Type))
		var u32 [4]byte
		binary.BigEndian.PutUint32(u32[:], uint32(len(a.Data)))
		buf = append(buf, u32[:]...)
		buf = append(buf, a.Data...)
	}
	return hex.EncodeToString(buf), nil
}

File: python/src/omne_sdk/abi.py
"""Omne ABI wire encoding for contract calls.

Port of sdk/typescript/src/contract.ts. Encodes a method call into the binary
wire format the runtime decodes, placed (hex-encoded) in a transaction's `data`
field or an omne_call `data` field:

    MAGIC "OMNE" (4) | VERSION 0x01 (1) | method_len u16 BE | method_utf8
        | arg_count u16 BE | [ type u8 | data_len u32 BE | data ]*

NOTE: ABI integers are BIG-endian here (the codec), distinct from the
LITTLE-endian transaction-hash preimage in transaction.py. Keep them separate.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum

from .address import parse_address

_ABI_MAGIC = b"OMNE"
_ABI_VERSION = 0x01
_MAX_METHOD_NAME_LEN = 256
_MAX_ARG_COUNT = 64


class ArgType(IntEnum):
    U32 = 0x01
    U64 = 0x02
    I32 = 0x03
    I64 = 0x04
    STRING = 0x05
    BYTES = 0x06
    BOOL = 0x07
    ADDRESS = 0x08
    U128 = 0x09


@dataclass(frozen=True)
class AbiArgument:
    type: ArgType
    data: bytes


class AbiEncode:
    """Convenience builders for typed ABI arguments (big-endian numerics)."""

    @staticmethod
    def u32(value: int) -> AbiArgument:
        return AbiArgument(ArgType.U32, int(value).to_bytes(4, "big"))

    @staticmethod
    def u64(value: int) -> AbiArgument:
        return AbiArgument(ArgType.U64, int(value).to_bytes(8, "big"))

    @staticmethod
    def i32(value: int) -> AbiArgument:
        return AbiArgument(ArgType.I32, int(value).to_bytes(4, "big", signed=True))

    @staticmethod
    def i64(value: int) -> AbiArgument:
        return AbiArgument(ArgType.I64, int(value).to_bytes(8, "big", signed=True))

    @staticmethod
    def u128(value: int) -> AbiArgument:
        return AbiArgument(ArgType.U128, int(value).to_bytes(16, "big"))

    @staticmethod
    def string(value: str) -> AbiArgument:
        return AbiArgument(ArgType.STRING, value.encode("utf-8"))

    @staticmethod
    def bytes_(value: bytes) -> AbiArgument:
        return AbiArgument(ArgType.BYTES, bytes(value))

    @staticmethod
    def bool_(value: bool) -> AbiArgument:
        return AbiArgument(ArgType.BOOL, b"\x01" if value else b"\x00")

    @staticmethod
    def address(omne_address: str) -> AbiArgument:
        payload = parse_address(omne_address)
        if len(payload) != 32:
            raise ValueError(f"Address must be 32 bytes, got {len(payload)}")
        return AbiArgument(ArgType.ADDRESS, payload)

    @staticmethod
    def address_bytes(payload: bytes) -> AbiArgument:
        if len(payload) != 32:
            raise ValueError(f"Address must be 32 bytes, got {len(payload)}")
        return AbiArgument(ArgType.ADDRESS, bytes(payload))


def encode_contract_call(method: str, args: list[AbiArgument] | None = None) -> str:
    """Encode a method call into the Omne ABI wire format; returns bare hex."""
    args = args or []
    method_bytes = method.encode("utf-8")
    if len(method_bytes) > _MAX_METHOD_NAME_LEN:
        raise ValueError(f"Method name exceeds {_MAX_METHOD_NAME_LEN} byte limit")
    if len(args) > _MAX_ARG_COUNT:
        raise ValueError(f"Argument count {len(args)} exceeds maximum of {_MAX_ARG_COUNT}")

    out = bytearray()
    out += _ABI_MAGIC
    out.append(_ABI_VERSION)
    out += len(method_bytes).to_bytes(2, "big")
    out += method_bytes
    out += len(args).to_bytes(2, "big")
    for arg in args:
        out.append(int(arg.type))
        out += len(arg.data).to_bytes(4, "big")
        out += arg.data
    return out.hex()
